107-tee-boot-attestation.md

```diff
@@ -0,0 +1,498 @@
+and make progress.
+-->
+ * TODO
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this enhancement is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+With TEE secure processors acting as the ultimate root-of-trust, much of the
+registration and attestation process's behavior should be identical to that of
+if vTPM devices were the root of trust. However, some registration steps will be
+handled before the a Keylime agent is initialized.
+
+It is critical to establish trust in a CVM as early as possible. As such,
+typical CVMs encrypt their root disks, and require a key stored on an
+attestation server to unlock the root disk and begin running any operating
+systems or applications on the CVM. To get this key from the server, a CVM would
+have to fetch its TEE evidence from the secure processor and send it as input to
+the server. If the server is able to validate the TEE evidence, it will then
+release the key. At this point, CVMs could decrypt their root disks and begin
+running their OS and applications trusting that they are in-fact running
+confidentially.
+
+[SVSM](https://github.com/coconut-svsm/svsm) is a firmware that runs in CVMs and
+is responsible for performing TEE attestation. SVSM performs TEE attestation to
+unlock the CVM's root disk and establish the initial vTPM state. SVSM will
+perform the initial communication with the Keylime Registrar.
+
+vTPMs will be present to ensure the integrity of software running on a CVM (i.e.
+runtime attestation of CVMs). Furthermore, initial vTPM state will be used to
+derive the key to unlock the CVM root disk. Specifically, encrypted VM images
+will be sealed against the initial vTPM state of the CVM. The initial vTPM state
+key will then be encrypted with a TEE Attestation Key stored in the Keylime
+Registrar. Upon CVM boot, the CVM will have access to the encrypted initial vTPM
+state key. To unlock the root disk, the CVM will need the decrypted initial vTPM
+state key. The only way for the CVM to decrypt its initial vTPM state is to
+attest its TEE evidence with the Registrar. Upon successful TEE attestation, the
+Registrar will decrypt the initial vTPM state and send it back to the CVM. With
+this, initial vTPM state is tied to TEE attestation, and initial vTPM state will
+need to be available to run any workloads on CVMs. Thus, TEE attestation will be
+required for running any OS/applications and initializing initial vTPM state on
+a CVM.
```

Comment:
Is the initial vTPM state of a VM equal to the last vTPM state before previous shutdown of the VM? The question is basically whether the 'state of the vTPM' you are talking about is also preserving vTPM state that was created during the previous run of a VM, such as NVRAM indices that may hold keys etc.

Reply:
Yes and no. Things like PCR state are not persistent, but other state such as NVRAM and seeds that are used to derive keys from vTPM are preserved.

I think @stefano-garzarella may be able to explain a bit more.

Reply:
Yes and no. Things like PCR state are not persistent, but other state such as NVRAM and seeds that are used to derive keys from vTPM are preserved.